### ISSUE TITLE
Consumes and produces media-types defined on Docket are incorrectly merged together with consumes/produces media-types that are defined on a resource level

### ISSUE THREAD

**Issue by mzeijen** [springfox/springfox#1916]
**Springfox version:** 2.7.0
**Issue type:** **Bug report.**
**Repo that reproduces the bug:** https://github.com/mzeijen/springfox-mediatype-override-bug
See the README.md for more info on how to reproduce it.

**Description:**

When you define the produces and/or consumes media-types on a Docket level then these media-types aren't overwritten by media-types defined on a resource level but they are merged together. 

The demo app demonstrates this behaviour, but below you can see the Swagger specification I get and what I expect when defining the produces and consumes on a Docket level and try to override them by setting the produces and consumes on the @RequestMapping annotation of the controller method.

[actual-swagger-spec.json](https://raw.githubusercontent.com/mzeijen/springfox-mediatype-override-bug/master/actual-swagger-spec.json)
[expected-swagger-spec.json](https://raw.githubusercontent.com/mzeijen/springfox-mediatype-override-bug/master/expected-swagger-spec.json)

**Possible problem location**:

The `springfox.documentation.spi.service.contexts.OperationContext` contains the methods `consumes` and `produces`. These use a `Sets.union` to merge the `requestContext` and `documentationContext` produces/consumes together. I believe this should not merge these together and only use the produces or consumes from the requestContext. 

The `MediaTypeReader` then needs to be fixed that it doesn't add the default values if document global consumes and produces are defined.

I will try this myself and will try to create a pull request for this.

**Comment by dilipkrish** [comment]
thank you for the detailed bug report! @mzeijen

**Comment by dilipkrish** [comment]
@mzeijen thank you for helping out with this issue! 🙇

**Comment by mzeijen** [comment]
Awesome that you merged the pull request. Thx.

When do you expect a new release?

**Comment by bjconlan-fc** [comment]
So how does one now use a 'merging' produces option? (global + method level)? This seems like a valid use case when using the content negotiating view resolver. Also when using 'application content types' ie:
```
@GetRequest(value = "test", produces = "application/vnd.ns.test.v1+json"}
Test getTest() throws Exception {
  throw new Exception("not an 'application/vnd.ns.test.v1'");
}
```
In this case spring correctly the produces and error payload in json with the content type "application/json" it would be good to have this recognized in the docs via the use of the 'merging global produces' docket builder.

**Comment by dilipkrish** [comment]
Dont think its related to this library as much as it is how spring handles the example usecae

**Comment by bjconlan-fc** [comment]
> Dont think its related to this library as much as it is how spring handles the example usecae

This is all regarding the swagger ui interface dropdown for requested 'accept-types'. (it should allow for merging a globally registered response type registered in the docket ie for a spring boot app:
```java
    // Spring webs default exception handlers
    // @see org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
    var globalResponses = Stream.of(
        HttpStatus.METHOD_NOT_ALLOWED,
        HttpStatus.UNSUPPORTED_MEDIA_TYPE,
        HttpStatus.NOT_ACCEPTABLE,
        HttpStatus.INTERNAL_SERVER_ERROR,
        HttpStatus.BAD_REQUEST,
        HttpStatus.NOT_FOUND,
        HttpStatus.SERVICE_UNAVAILABLE)
        .map(httpStatus -> new ResponseMessageBuilder()
            .code(httpStatus.value())
            .message(httpStatus.getReasonPhrase())
            .responseModel(errorResponse)
            .build())
        .collect(Collectors.toList());

    // TODO register global producer for 'application/json' instead of having them explicitly declared on all handlers
    //      https://github.com/springfox/springfox/issues/1916
    return new Docket(DocumentationType.SWAGGER_2)
        // Include global result types (@ControllerAdvice changes need to be propagated here).
        .additionalModels(typeResolver.resolve(ErrorResponse.class))
        .genericModelSubstitutes(Mono.class, Flux.class)
        .globalResponseMessage(RequestMethod.HEAD, globalResponses)
        .globalResponseMessage(RequestMethod.GET, globalResponses)
        .globalResponseMessage(RequestMethod.PUT, globalResponses)
        .globalResponseMessage(RequestMethod.POST, globalResponses)
        .globalResponseMessage(RequestMethod.PATCH, globalResponses)
        .globalResponseMessage(RequestMethod.DELETE, globalResponses)
...
```
This should merge both globally registered and locally specified 'produce' types (or provide the option to explicitly merge or override)

**Comment by dilipkrish** [comment]
@bjconlan-fc I don't follow your example. There is no media type specified to influence `produces`. The docket allows for you to specify a global set via `docket.produces` and individual APIs can specify specific media types and it will be merged

**Comment by bjconlan-fc** [comment]
@dilipkrish the `.produces(Sets.newHashSet("application/json"))` was ommited from the example because it does nothing (ie doesn't merge just becomes overwritten since said bug was 'fixed'). So if I were to introduce it an annotated the appropriate controller method with a produces this 'global' produces isn't merged it gets overwritten (so the workaround is to explicitly specify both on the handler method which isn't exactly the correct place to handle random side effects that happen at a global level.

Basically the use case Identified by this ticket isn't the use case used in my case. And perhaps a strategy for merging over overwriting globally registered 'produces' should be facilitated.
